proposals/p2511.md

```diff
@@ -0,0 +1,190 @@
+        including implicitly by `addr` self parameters, or at assignment
+        statements, making it easier to reason about where the value of a
+        variable can change in the control flow of a function.
+-   [Software and language evolution](/docs/project/goals.md#software-and-language-evolution)
+    -   This approach is conservative and can evolve to support assignment as a
+        subexpression.
+-   [Code that is easy to read, understand, and write](/docs/project/goals.md#code-that-is-easy-to-read-understand-and-write)
+    -   Assignments in subexpressions tend to be hard for humans to read and
+        understand. Disallowing them at the language level avoids the potential
+        for confusion, making code easier to read at the cost of making certain
+        constructs such as `n = arr[i++];` a little harder to write.
+    -   Easier to write a complete operator set because both `$` and `$=` can be
+        provided by implementing a single interface.
+-   [Practical safety and testing mechanisms](/docs/project/goals.md#practical-safety-and-testing-mechanisms)
+    -   Replaces correctness warning on `if (a = b)` with a language rule.
+-   [Interoperability with and migration from existing C++ code](/docs/project/goals.md#interoperability-with-and-migration-from-existing-c-code)
+    -   Providing largely the same set of symbols makes migration easier.
+    -   Some cost may be imposed by forcing a translation from assignment in
+        subexpressions to assignment as separate statements.
+
+## Alternatives considered
+
+### Allow assignment as a subexpression
+
+We could allow some or all forms of assignment as subexpressions, either with
+the same syntax or with some other syntax. For now, we are choosing to not
+pursue these options in order to gather data around how important they are. If
+we wish to allow assignment as a subexpression, we should consider adopting
+something like Python's "walrus operator" `variable := value` instead of using
+an `=` symbol.
+
+If we allow a three-term form of `for` statements, `for (init; cond; incr)`, we
+could allow assignment in the `incr` term. However, we currently do not support
+this syntax.
+
+### Allow chained assignment
+
+As a special case of assignment as a subexpression, we could allow chained
+assignment:
+
+```
+a = b = c = 0;
+```
+
+We could restrict this to only apply to simple assignment, not cases like
+`a += b -= c *= 2;`
+
+In leads issue [#451](https://github.com/carbon-language/carbon-lang/issues/451)
+it was decided that we would initially not support chained assignment.
+
+### Do not provide increment and decrement
+
+We could remove `++var;` and `--var;` in favor of `var += 1;` and `var -= 1;`.
+However, developers coming from C-family languages will expect these operators
+to exist, and they may more directly convey the intended semantics of counting
+than a `+=` operation would.
+
+### Define `$=` in terms of `$`
```

Comment:
FWIW, I suspect this will be a pain point for folks. I'm guessing there will be a lot of types where the potential costs are either immaterial or actually don't exist and the conceptual model will be substantially cleaner the other way.

Another thing to think about is how a generic function needs to be constrained. It would seem much more obvious to teach that you require `Add & Mul & Assign` and get `+`, `*`, `=`, `+=`, `*=`. The other way around seems likely to be much less obvious.

I wonder if this should even override the concern around performance.

Reply:
I may be confused, but doesn't the "Default" section of the design address this? (though, this alternative probably should mention the default interfaces)

Reply:
@jonmeow My understanding is that @chandlerc would like to get `$=` defined in terms of `$`, whereas this proposal previously did the opposite: it defined `$` in terms of `$=`.

@chandlerc I think you've made a good point here, and in fact I'm not sure the performance argument actually holds up. I can easily think of situations where for performance you wouldn't want either operation defined in terms of the other one. Defining `$=` in terms of `$` and `=` seems to be strictly better if we set that aside. I've made that change.